Add support for multipart/related

bluk opened the pull request:
- Use `multipart/related` but send as `multipart/form-data`
- For JSON data, send as a fake file that has a Content-Type
`application/json`
- Order the parts so real files are added last
- Add support for image/gif

Commit:
Add support for multipart/related

- Use `multipart/related` but send as `multipart/form-data`
- For JSON data, send as a fake file that has a Content-Type
`application/json`
- Order the parts so real files are added last
- Add support for image/gif

sdcoffey commented on spec/unit/http_client_spec.js:
```diff
@@ -226,6 +226,82 @@ describe('HttpClient', function () {
+          'Content-Type': 'multipart/related'
+        },
+        partHeaders: {
+          input_field_name: {
```
Is there a requirement for individual parts to have their own set of headers? If so, is there any way we can encapsulate them into the part itself? 

I'm thinking about something like: 

```js
let req = {
  verb: 'POST',
  path: '/',
  headers: {},
  body: {
    file_part: fs.createReadStream('...'),
    form_part: 'some-value',
    other_form_part: new FormPart({
      whatever_key: 'some-value',
      headers: {
        'Content-Type': 'application/json'
      }
    })
  }
};
```

sdcoffey replied:
I suppose this applies mostly for the content type of a given object in a body, but the point still stands.

sdcoffey commented on lib/braintreehttp/encoder.js:
```diff
@@ -3,13 +3,14 @@
 
   constructor() {
-    this._encoders = [new Json(), new Text(), new Multipart(), new FormEncoded()];
+    this._encoders = [new Json(), new Text(), new MultipartRelated(), new Multipart(), new FormEncoded()];
```
It seems like there's a lot of code duplication between the `multipart` and `mutlipart_related`. Is there any way we can generalize some of the logic or even make one work for both content types?

Commit:
Introduce FormPart class to organize individual form parts

- addresses
https://github.com/braintree/braintreehttp_node/pull/2#discussion_r177256550

Commit:
Integrate multipart/related into multipart serializer

jackellenberger commented:
test

sdcoffey commented on lib/braintreehttp/serializer/multipart.js:
```diff
@@ -40,22 +43,38 @@ class Multipart {
+    }
+
+    if (formPartContentType === 'application/json') {
+      contentBuffer = Buffer.from(JSON.stringify(formPartValue) + Multipart._CRLF)
```
I think this is fine for now, but we should be prepared to support other content types in the future.

sdcoffey replied:
It's a bit of a hack, but you could construct a new encoder here and pass the part directly in. The encoder will handle serializing the data since there's already a `header` property with `Content-Type` set.

bluk replied:
We had to break a cyclical dependency in the `Reuse Encoder in Multipart`. Let us know what you think. It does allow us to break the dependency between Encoder and serializers as well to allow specific serializers to be used per Encoder.

sdcoffey reviewed (approved):
Just one comment about future-proofing this work, but otherwise I think it looks good!

Commit:
fix linting errors

Commit:
Reuse Encoder in Multipart

- Have Encoder take a list of serializers so that it
  breaks a cycical dependency as well as allows a specific
  set of serializers to be used

Commit:
Export FormPart from the main braintreehttp.js

demerino reviewed (approved):
We do an end-to-end test yet or does this need merging first?

jackellenberger commented:
@demerino we cooked up a lil test that we didn't commit to hit the live endpoint with some of Bryant's sandbox disputes and it seemed to work!

Commit:
Fix FormPart headers initialization
